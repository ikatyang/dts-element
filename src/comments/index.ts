export {
  IJSDocComment,
  IJSDocCommentOptions,
  create_jsdoc_comment,
} from './jsdoc-comment';
export {
  IMultiLineComment,
  IMultiLineCommentOptions,
  create_multi_line_comment,
} from './multi-line-comment';
export {
  ISingleLineComment,
  ISingleLineCommentOptions,
  create_single_line_comment,
} from './single-line-comment';
