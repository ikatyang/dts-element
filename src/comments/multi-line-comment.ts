import * as ts from 'typescript';
import { ElementKind } from '../constants';
import { create_element, IElement, IElementOptions } from '../element';

export interface IMultiLineCommentOptions extends IElementOptions {
  text: string;
  align?: boolean;
  prefix?: string;
}

export interface IMultiLineComment
  extends IElement<ElementKind.MultiLineComment>,
    IMultiLineCommentOptions {}

export const create_multi_line_comment = (
  options: IMultiLineCommentOptions,
): IMultiLineComment => ({
  ...create_element(ElementKind.MultiLineComment),
  ...options,
});

/**
 * @hidden
 */
export interface IAddMultiLineCommentOptions {
  method?:
    | typeof ts.addSyntheticLeadingComment
    | typeof ts.addSyntheticTrailingComment;
  trailing_new_line?: boolean;
}

/**
 * @hidden
 */
// tslint:disable:ter-indent
export const add_multi_line_comment = <T extends ts.Node>(
  node: T,
  element: IMultiLineComment,
  options: IAddMultiLineCommentOptions = {},
): T =>
  (options.method || ts.addSyntheticLeadingComment)(
    /* node                */ node,
    /* kind                */ ts.SyntaxKind.MultiLineCommentTrivia,
    /* text                */ element.text
      .replace(/^/gm, element.prefix === undefined ? '' : element.prefix)
      .replace(
        /\n|$/g,
        match =>
          element.align === true ? (match === '\n' ? '\n *' : '\n ') : match,
      ),
    /* hasTrailingNewLine  */ options.trailing_new_line === undefined
      ? true
      : options.trailing_new_line,
  );
// tslint:enable:ter-indent

/**
 * @hidden
 */
export const transform_multi_line_comment = (
  element: IMultiLineComment,
  _path: IElement<any>[],
) =>
  add_multi_line_comment(ts.createOmittedExpression(), element, {
    trailing_new_line: false,
    method: ts.addSyntheticTrailingComment,
  });
