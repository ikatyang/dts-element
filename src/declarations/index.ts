// tslint:disable:max-line-length

export * from './class-declaration';
export * from './enum-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './global-declaration';
export * from './interface-declaration';
export * from './module-declaration';
export * from './namespace-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
