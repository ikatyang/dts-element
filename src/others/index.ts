// tslint:disable:max-line-length

export {
  IIndexSignature,
  IIndexSignatureOptions,
  create_index_signature,
} from './index-signature';
export {
  ITopLevelElement,
  ITopLevelElementOptions,
  create_top_level_element,
} from './top-level-element';
export {
  ITripleSlashReference,
  ITripleSlashReferenceOptions,
  create_triple_slash_reference,
} from './triple-slash-reference';
export {
  ITypePredicate,
  ITypePredicateOptions,
  create_type_predicate,
} from './type-predicate';
