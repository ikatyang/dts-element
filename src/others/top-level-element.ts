import * as ts from 'typescript';
import { ITopLevelMember } from '../collections';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';

export interface ITopLevelElementOptions extends IElementOptions {
  members: ITopLevelMember[];
}

export interface ITopLevelElement
  extends IElement<ElementKind.TopLevelElement>,
    ITopLevelElementOptions {}

export const create_top_level_element = (
  options: ITopLevelElementOptions,
): ITopLevelElement => ({
  ...create_element(ElementKind.TopLevelElement),
  ...options,
});

export const is_top_level_element = (value: any): value is ITopLevelElement =>
  is_element(value) && value.kind === ElementKind.TopLevelElement;

/**
 * @hidden
 */
export const transform_top_level_element = (
  element: ITopLevelElement,
  path: IElement<any>[],
) => {
  const source_file = ts.createSourceFile(
    /* fileName        */ '',
    /* sourceText      */ '',
    /* languageVersion */ ts.ScriptTarget.Latest,
    /* setParentNodes  */ false,
    /* scriptKind      */ ts.ScriptKind.TS,
  );

  source_file.statements = ts.createNodeArray(
    element.members.map(member => transform(member, path) as ts.Statement),
  );
  return source_file;
};
