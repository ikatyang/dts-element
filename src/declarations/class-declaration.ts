import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { IClassMember } from '../members';
import { IIndexSignature } from '../others/index-signature';
import { transform } from '../transform';
import { IGeneralType } from '../types/general-type';
import {
  add_declare_modifier_if_need,
  create_expression_for_general_type,
  create_type_parameters,
} from '../utils';
import { IGenericDeclaration } from './generic-declaration';

export interface IClassDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  members?: (IClassMember | IIndexSignature)[];
  abstract?: boolean;
  extends?: IGeneralType;
}

export interface IClassDeclaration
  extends IElement<ElementKind.ClassDeclaration>,
    IClassDeclarationOptions {}

export const create_class_declaration = (
  options: IClassDeclarationOptions,
): IClassDeclaration => ({
  ...create_element(ElementKind.ClassDeclaration),
  ...options,
});

export const is_class_declaration = (value: any): value is IClassDeclaration =>
  is_element(value) && value.kind === ElementKind.ClassDeclaration;

// tslint:disable:ter-indent max-line-length

/**
 * @hidden
 */
export const transform_class_declaration = (
  element: IClassDeclaration,
  path: IElement<any>[],
) =>
  ts.createClassDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
      [
        ...(element.export === true
          ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
          : []),
        ...(element.abstract === true
          ? [ts.createToken(ts.SyntaxKind.AbstractKeyword)]
          : []),
      ],
      path,
    ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ element.extends === undefined
      ? []
      : [
          ts.createHeritageClause(
            /* token */ ts.SyntaxKind.ExtendsKeyword,
            /* types */ [
              create_expression_for_general_type(element.extends, path),
            ],
          ),
        ],
    /* members         */ (element.members || [])
      .map(member => transform(member, path) as ts.ClassElement),
  );
