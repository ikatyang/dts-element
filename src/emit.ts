import * as ts from 'typescript';
import { ITopLevelMember } from './collections';
import { IElement } from './element';
import { create_top_level_element } from './others/top-level-element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (
  element: IElement<any> | ITopLevelMember[],
  options?: ts.PrinterOptions,
) => {
  const target = Array.isArray(element)
    ? create_top_level_element({ members: element })
    : element;
  return emit_native(transform(target), options);
};
