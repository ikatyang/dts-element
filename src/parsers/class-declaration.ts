import * as ts from 'typescript';
import {
  create_class_declaration,
  IClassDeclaration,
} from '../declarations/class-declaration';
import { IGenericDeclaration } from '../declarations/generic-declaration';
import { IClassMember } from '../members/class-member';
import { parse_native } from '../parse';
import { IGeneralType } from '../types/general-type';
import { has_kind, if_defined } from '../utils';
import { parse_identifier } from './identifier';

export const parse_class_declaration = (
  node: ts.ClassDeclaration,
): IClassDeclaration =>
  create_class_declaration({
    name: parse_identifier(node.name as ts.Identifier),
    export: has_kind(node.modifiers, ts.SyntaxKind.ExportKeyword),
    abstract: has_kind(node.modifiers, ts.SyntaxKind.AbstractKeyword),
    generics: if_defined(
      node.typeParameters,
      generics => generics.map(parse_native) as IGenericDeclaration[],
    ),
    extends: if_defined(
      node.heritageClauses,
      heritage_clauses =>
        parse_native(heritage_clauses[0].types[0]) as IGeneralType,
    ),
    members: node.members.map(parse_native) as IClassMember[],
  });
