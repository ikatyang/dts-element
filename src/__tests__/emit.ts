import * as ts from 'typescript';
import { create_multi_line_comment } from '../comments/multi-line-comment';
import { create_class_declaration } from '../declarations/class-declaration';
import { emit } from '../emit';
import { create_top_level_element } from '../others/top-level-element';
import { create_general_type } from '../types/general-type';

const top_level_element = create_top_level_element({
  members: [
    create_multi_line_comment({
      text: `hello\n\nworld`,
      align: true,
      prefix: '~ ',
    }),
    create_class_declaration({
      name: 'AAA',
      extends: create_general_type({
        name: 'SomethingElse',
      }),
    }),
  ],
});

it('should return correctly', () => {
  expect(JSON.stringify(emit(top_level_element))).toMatchSnapshot();
});

it('should return correctly with element = array', () => {
  expect(JSON.stringify(emit(top_level_element.members))).toMatchSnapshot();
});

it('should return correctly with removeComments', () => {
  expect(
    JSON.stringify(
      emit(top_level_element, {
        removeComments: true,
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with newLine', () => {
  expect(
    JSON.stringify(
      emit(top_level_element, {
        newLine: ts.NewLineKind.CarriageReturnLineFeed,
      }),
    ),
  ).toMatchSnapshot();
});

it('should return correctly with removeComments and newLine', () => {
  expect(
    JSON.stringify(
      emit(top_level_element, {
        removeComments: true,
        newLine: ts.NewLineKind.CarriageReturnLineFeed,
      }),
    ),
  ).toMatchSnapshot();
});
