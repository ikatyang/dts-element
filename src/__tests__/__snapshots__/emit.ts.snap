// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `"\\"/*~ hello\\\\n *~\\\\n *~ world\\\\n */\\\\ndeclare class AAA extends SomethingElse {\\\\n}\\\\n\\""`;

exports[`should return correctly with element = array 1`] = `"\\"/*~ hello\\\\n *~\\\\n *~ world\\\\n */\\\\ndeclare class AAA extends SomethingElse {\\\\n}\\\\n\\""`;

exports[`should return correctly with newLine 1`] = `"\\"/*~ hello\\\\r\\\\n *~\\\\r\\\\n *~ world\\\\r\\\\n */\\\\r\\\\ndeclare class AAA extends SomethingElse {\\\\r\\\\n}\\\\r\\\\n\\""`;

exports[`should return correctly with removeComments 1`] = `"\\"declare class AAA extends SomethingElse {\\\\n}\\\\n\\""`;

exports[`should return correctly with removeComments and newLine 1`] = `"\\"declare class AAA extends SomethingElse {\\\\r\\\\n}\\\\r\\\\n\\""`;
